### PR TITLE
Timeline utils for fragments

### PR DESCRIPTION
**Pull request by Voycawojka**
## Description

Utils for fragments' timelines

## Motivation and Context

https://github.com/Knotx/knotx-fragments-chrome-extension/issues/24

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: dist/index.html [hunk 1]
```diff
@@ -19,6 +19,10 @@
 <html>
   <head>
     <title>Knot.x Chrome Devtool Extension</title>
+
+    <!-- Workaround for a vis.js bug. Without manually importing the CSS timeline throws JS errors and doesn't work -->
+    <!-- https://github.com/almende/vis/issues/2739 -->
+    <link href="https://unpkg.com/vis-timeline@7.1.2/dist/vis-timeline-graph2d.min.css" rel="stylesheet">
```
Comment (SzajekSzymon) [inline dist/index.html]: Please use : import 'vis/dist/vis-timeline-graph2d.min.css'; in js file

File: src/js/helpers/timeline/declarationHelper-transform.spec.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  };
+
+  const timeline = constructTimeline(mock.singleNode);
+
+  expect(timeline.items.get()).toStrictEqual(expectedTimeline.items);
+  expect(timeline.groups.get()).toStrictEqual(expectedTimeline.groups);
+});
+
+test('Group order is correct', () => {
+  const timeline = constructTimeline(mock.sinleNodeWithMultipleChildrenLevels);
```
Comment (SzajekSzymon) [inline src/js/helpers/timeline/declarationHelper-transform.spec.js]: typo in sinleNodeWithMultipleChildrenLevels

**Comment by SzajekSzymon** [comment]
Add missing license pls.

File: src/js/devtools/App.jsx [hunk 3]
```diff
@@ -16,25 +16,28 @@
...
+const App = (/*{ tabId }*/) => {
   const { themeName: chromeTheme } = chrome.devtools.panels;
   const theme = chromeTheme === 'default' ? defaultTheme : darkTheme;
 
   return (
     <ThemeProvider theme={theme}>
       <div className="app">
-        <FragmentList tabId={tabId} />
+        <ExampleTimeline />
+        {/* <FragmentList tabId={tabId} /> */}
```
Comment (SzajekSzymon) [inline src/js/devtools/App.jsx]: Please uncomment this. Bug has been fixed.

File: src/js/devtools/App.jsx [hunk 4]
```diff
@@ -16,25 +16,28 @@
 
 import React from 'react';
 import { ThemeProvider } from 'styled-components';
-import PropTypes from 'prop-types';
-import FragmentList from '../components/FragmentList/fragmentList';
+// import PropTypes from 'prop-types';
```
Comment (tomaszmichalak) [inline src/js/devtools/App.jsx]: Why those are commented?

File: src/js/devtools/App.jsx [hunk 5]
```diff
@@ -16,25 +16,28 @@
...
+// import FragmentList from '../components/FragmentList/fragmentList';
 import { defaultTheme, darkTheme } from './themes';
+import ExampleTimeline from '../components/Timeline/exampleTimeline';
```
Comment (tomaszmichalak) [inline src/js/devtools/App.jsx]: Please move it after the `FragmentList` declaration.

File: src/js/devtools/App.jsx [hunk 6]
```diff
@@ -16,25 +16,28 @@
...
 
-const App = ({ tabId }) => {
+// eslint-disable-next-line spaced-comment
+const App = (/*{ tabId }*/) => {
```
Comment (tomaszmichalak) [inline src/js/devtools/App.jsx]: Uncomment this line please.

File: src/js/devtools/App.jsx [hunk 7]
```diff
@@ -16,25 +16,28 @@
...
+        {/* <FragmentList tabId={tabId} /> */}
       </div>
     </ThemeProvider>
   );
 };
 
-App.propTypes = {
-  tabId: PropTypes.number.isRequired,
-};
+// App.propTypes = {
```
Comment (tomaszmichalak) [inline src/js/devtools/App.jsx]: Should be commented?

File: src/js/devtools/App.jsx (hunk 4 above)
Comment (Voycawojka) [reply src/js/devtools/App.jsx]: It was just for testing purposes, I commited it this way by mistake

File: src/js/helpers/timeline/declarationHelper.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    total[subgroupOf] = [...total[subgroupOf], group];
+  }
+
+  total[group] = total[group] || [];
+
+  return total;
+};
+
+export const extractGroupData = (items) => Object.entries(items
+  .map(({ group, subgroupOf }) => ({ group, subgroupOf }))
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper.js]: What does it do?

Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper.js]: It does practically nothing, must have slipped under my radar ;)

File: src/js/helpers/timeline/declarationHelper.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  ? [root, ...getProcessedNodes(getNextNodes(root))]
+  : [root]);
+
+const generateGroupName = (node) => node.label;
+
+const generateSubgroupData = (parent) => (parent ? { subgroupOf: generateGroupName(parent) } : {});
+
+const createTimelineItem = (node, parent) => ({
+  id: node.id,
+  start: parseInt(node.started, 10),
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper.js]: What is 10 - maybe some constant.

File: src/js/helpers/timeline/declarationHelper.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  end: parseInt(node.finished, 10),
+  content: '',
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper.js]: `constant EMPTY_LABEL = '';` ?

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    response: {
+      invocations: {},
+      transition: '_success',
+    },
+  };
+
+  expect(getProcessedNodes(graph)).toStrictEqual([graph, child]);
+});
+
+test('Unpocessed child nodes are not returned', () => {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]:
```suggestion
test('Unprocessed child nodes are not returned', () => {
```

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getProcessedNodes } from './declarationHelper';
+
+test('One-node-graph is returned as is', () => {
+  const graph = {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]: `graph` -> `rootNode` ?

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js (hunk 11 above)
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]: What will happen when the root node is `unprocessed` or even `empty`?

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    response: {
+      invocations: {},
+      transition: '_error',
+    },
+  };
+
+  expect(getProcessedNodes(graph)).toStrictEqual([graph]);
+});
+
+test('Composites are treated as a whole', () => {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]: Maybe add simple test scenario with composite as a root node?

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        label: 'A label',
+        status: 'success',
+        type: 'single',
+        response: {
+          invocations: {},
+          transition: '_success',
+        },
+      },
+    ],
+    response: {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]: Please add case when composite defines next node.

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+      invocations: {},
+      transition: '_success',
+    },
+  };
+
+  expect(getProcessedNodes(graph)).toStrictEqual([graph, compositeChild]);
+});
+
+test('Multiple transition are parsed correctly', () => {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-nodes.spec.js]: Move it to the beginning please.

File: src/js/helpers/timeline/declarationHelper-groups.spec.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { extractGroupData, reduceSubgroups } from './declarationHelper';
+
+test('Group with parent information is correctly transform into group with subgroups information', () => {
+  const alreadyReducedData = { parent: [] };
+  const currentData = { group: 'child', subgroupOf: 'parent' };
+  const expectedData = { parent: ['child'], child: [] };
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-groups.spec.js]: You have `'child'` and and `child` in the same line - can be misleading. The same with parent with `parent`.

How about something like that:
```
  const parentGroupName = 'parentGroup';
  const childGroupName = 'childGroup'
  const alreadyReducedData = { parent: [] };
  const currentData = { group: childGroupName, subgroupOf: parentGroupName };
  const expectedData = { parent: [childGroupName], child: [] };
```

File: src/js/helpers/timeline/declarationHelper-groups.spec.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { extractGroupData, reduceSubgroups } from './declarationHelper';
+
+test('Group with parent information is correctly transform into group with subgroups information', () => {
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-groups.spec.js]: Please add a case when `alreadyReducedData` contains some data.

File: src/js/helpers/timeline/declarationHelper-groups.spec.js [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  ];
+
+  expect(extractGroupData(groupItems)).toStrictEqual(expectedGroupData);
+});
+
+test('Flatten subgroups and keep ids in parent', () => {
+  const groupItems = [
+    { group: 'A' },
+    { group: 'B' },
+    { group: 'A-1', subgroupOf: 'A' },
```
Comment (tomaszmichalak) [inline src/js/helpers/timeline/declarationHelper-groups.spec.js]: Does the order matter? What if the item with `subgroupOf: 'A'` appears before item with `group: 'A'`?

File: src/js/helpers/timeline/declarationHelper-nodes.spec.js (hunk 11 above)
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper-nodes.spec.js]: It'll return an empty array and therefore the timeline will be empty. I can cover this case, but is it even possible for a root node to be unprocessed?

File: src/js/helpers/timeline/declarationHelper-items.spec.js [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+import { createTimelineItems } from './declarationHelper';
+import * as mock from './declerationHelper.mock';
+
+test('Single node generates one timeline item', () => {
+  const expectedItem = {
+    id: 'A',
+    start: 12345,
+    end: 54321,
+    content: '',
```
[the remark this answers is not in the document]
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper-items.spec.js]: It's a label, but there's not much I we can do about this name, because it's a property required by Vis

File: src/js/helpers/timeline/declarationHelper-items.spec.js [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+ * limitations under the License.
+ */
+
+import { createTimelineItems } from './declarationHelper';
+import * as mock from './declerationHelper.mock';
+
+test('Single node generates one timeline item', () => {
+  const expectedItem = {
+    id: 'A',
+    start: 12345,
```
[the remark this answers is not in the document]
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper-items.spec.js]: `start` and `end` are required by Vis, so we can't change it here

File: src/js/helpers/timeline/declarationHelper-groups.spec.js (hunk 16 above)
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper-groups.spec.js]: It cannot be the way you suggested, but right - it is misleading. `'child'` + `child` and `'parent'` + `parent` have the same name because they represent the same groups and are arbitrary.  I'll change those names to something like this to make it clearer:

```
const alreadyReducedData = { parentName: [] };
const currentData = { group: 'childName', subgroupOf: 'parentName' };
const expectedData = { parentName: ['childName'], childName: [] };
```

File: src/js/helpers/timeline/declarationHelper-groups.spec.js (hunk 18 above)
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper-groups.spec.js]: It matters, this is the order in which the groups will be displayed in the timeline (first one on top)

File: src/js/helpers/timeline/declarationHelper.js [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+    total[subgroupOf] = [...total[subgroupOf], group];
+  }
+
+  total[group] = total[group] || [];
+
+  return total;
+};
+
+export const extractGroupData = (items) => Object.entries(items
+  .reduce(reduceSubgroups, {}))
```
Comment (MichalGraziowskiCognifide) [inline src/js/helpers/timeline/declarationHelper.js]: How about putting it in one line? On the first glance I thought that the `map` below is to the direct result of this `reduce` and I was confused.

File: src/js/helpers/timeline/declarationHelper.js [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+  nestedGroups: subgroups,
+});
+
+export const constructTimeline = (json) => {
+  const items = getProcessedNodes(json)
+    .flatMap((node) => createTimelineItems(node));
+
+  const groups = extractGroupData(items)
+    .map((group) => (group.subgroups.length > 0 ? group : { ...group, subgroups: null }))
+    .map((group, index) => createTimelineGroup(group.name, index, group.subgroups));
```
Comment (MichalGraziowskiCognifide) [inline src/js/helpers/timeline/declarationHelper.js]: I believe it could be simplified to:
```
.map(({name, subgroups}, index) => createTimelineGroup(name, index, subgroups. length ? subgroups : null));
```

File: src/js/helpers/timeline/declarationHelper.js [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    createTimelineItem(node, parent),
+    ...node.subtasks
+      .flatMap((subtask) => getProcessedNodes(subtask))
+      .flatMap((subtask) => createTimelineItems(subtask, node))]
+  : [
+    createTimelineItem(node, parent),
+  ]);
+
+export const reduceSubgroups = (prev, { group, subgroupOf }) => {
+  const total = { ...prev };
```
Comment (MichalGraziowskiCognifide) [inline src/js/helpers/timeline/declarationHelper.js]: Spread operator creates only shallow copy, it's not the problem, right?

File: src/js/helpers/timeline/declarationHelper.js [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+};
+
+export const getGraphWithUniqueLabels = (root, takenLabels = {}) => ({
+  ...root,
+  uniqueLabel: generateUniqueLabel(root.label, takenLabels),
+  ...(root.on
+    ? {
+      on: Object.entries(root.on)
+        .map(([transition, node]) => [transition, getGraphWithUniqueLabels(node, takenLabels)])
+        .reduce((total, [transition, node]) => ({ ...total, [transition]: node }), {}),
```
Comment (MichalGraziowskiCognifide) [inline src/js/helpers/timeline/declarationHelper.js]: https://www.richsnapp.com/blog/2019/06-09-reduce-spread-anti-pattern - I'll just leave it here.
tl;dr: reduce with spread operator has almost O(n^2) complexity.
I don't know if it's an issue for you.

File: src/js/helpers/timeline/declarationHelper.js (hunk 22 above)
Comment (tomaszmichalak) [reply src/js/helpers/timeline/declarationHelper.js]: Nice!

**Comment by tomaszmichalak** [comment]
@Voycawojka This PR is ready to merge. Please validate last two comments.

File: src/js/helpers/timeline/declarationHelper.js (hunk 23 above)
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper.js]: I know, shallow copy is enough

File: src/js/helpers/timeline/declarationHelper.js (hunk 24 above)
Comment (Voycawojka) [reply src/js/helpers/timeline/declarationHelper.js]: Good too know, I didn't realize that. But I think it's ok here, especially since an alternative would be to mutate a function parameter which is forbidden by our eslint and since most of the data will be fairly small so there won't be much difference anyway

Comment (MichalGraziowskiCognifide) [reply src/js/helpers/timeline/declarationHelper.js]: Sure, no problem 😄
